vllm/engine/multiprocessing/client.py

```diff
@@ -203,8 +205,23 @@ async def run_output_handler_loop(self):
                         self._errored_with = exception
 
                     if request_id is None:
+
                         for queue_i in tuple(self.output_queues.values()):
-                            queue_i.put_nowait(exception)
+
+                            msg = str("A batch generation failed. Inspect the "
+                                      "stacktrace to find the original error: "
+                                      f"{repr(exception)}")
+                            # If it is a runtime exception, we assume that
+                            # the engine is already dead, let's pass this
```

Comment:
Ah I think maybe I see- we're trying to send `MQEngineDeadError` instead of `MQEngineBatchError` if the engine is already dead... 🤔 

I'm not sure that checking for `RuntimeError` is the most robust, we should be able to instead check `if self.errored`

Reply:
Yeah... I broke a test  before because of this RuntimeError. There is already an exception handler that expect a RuntimeError and it will  shutdown the server next (In this case it already keep the behavior that we expect in this PR: without duplicated error stacktrace).

```python
# repo/vllm/entrypoints/launcher.py
@app.exception_handler(RuntimeError)
    async def runtime_error_handler(request: Request, __):
        """On generic runtime error, check to see if the engine has died.
        It probably has, in which case the server will no longer be able to
        handle requests. Trigger a graceful shutdown with a SIGTERM."""
        engine = request.app.state.engine_client
        if (not envs.VLLM_KEEP_ALIVE_ON_ENGINE_DEATH and engine.errored
                and not engine.is_running):
            logger.fatal("AsyncLLMEngine has failed, terminating server "
                         "process")
            # See discussions here on shutting down a uvicorn server
            # https://github.com/encode/uvicorn/discussions/1103
            # In this case we cannot await the server shutdown here because
            # this handler must first return to close the connection for
            # this request.
            server.should_exit = True
```

So, in this PR one of my contribution was to add a flow that handle Exception that are not RuntimeErrors, which cause the log with the stacktrace printed multiple times.

Reply:
So, the intent behind the extra `RuntimeError` there was to try to catch anything that might have killed the engine, but did not raise an `EngineDeadError`, which ideally should not happen. It's kinda equivalent to
```
try:
    handle_request(req)
except EngineDeadError:
    engine_dead_handler(req)
except RuntimeError:
    runtime_error_handler(req)
```
Maybe that should actually be changed to `@app.exception_handler(Exception)` to catch everything? Would that solve this problem entirely by just doing that? I had originally written these handlers, and I think I may have just made a bad assumption that unexpected things would be `RuntimeError`s, which is probably often true but not always true. (It's not true for your test case where you explicitly raised `Exception` 😄 ) 


Then regardless of the server handling, I do think the error handling here in the engine should be more robust than just checking for `RuntimeError`. If we want special logic based on whether the engine is already dead, then I think we can do:
```
batch_error = MQEngineDeadError(msg) if self.errored else MQEngineBatchError(msg)
```

Reply:
> Maybe that should actually be changed to @app.exception_handler(Exception) to catch everything?

Sort of, I added a custom for `MQEngineBatchError` because we are sure that the engine had an error and it was logged before. But, what if we got an unhandled exception not related to the engine? I think in those case let the exception go up to the server and let FastAPI/Uvicorn do their thing.

> batch_error = MQEngineDeadError(msg) if self.errored else MQEngineBatchError(msg)

Today, in practice any exception will eventually raise a MQEngineDeadError.
```python
    # Set to error state only on engine critical error
    # (and record only the first one)
    if is_engine_errored and not self._errored_with:
        self._errored_with = exception
```

Honestly, I don't know the best solution, I wrote my solution based on what I read the code and tried to keep the old the behavior. I found two scenarios: i) runtime exception, shutdown server immediately; ii) other exceptions, set the engine erroed, when another request arrive then throws MQEngineDeadError and shutdown the server next. A simplification would be treat everything as runtime errors... 

I introduced MQEngineBatchError thinking that in the future vLLM would be more robust to those errors, raise an error and keep up and running, but maybe it is a too soon feature. I would like hear more opinions on that.

Reply:
Thanks for all the clarification @wallashss! This is definitely a bit confusing to hold it all in my head 😅 

> But, what if we got an unhandled exception not related to the engine? I think in those case let the exception go up to the server and let FastAPI/Uvicorn do their thing.

That's a good distinction, yeah. I'm okay with leaving stack traces in for things that we don't think were caused by an exception that killed the engine.

>  ii) other exceptions, set the engine erroed, when another request arrive then throws MQEngineDeadError and shutdown the server next. A simplification would be treat everything as runtime errors...

Oh, so actually I don't think we want to wait for the second request to come in to throw `MQEngineDeadError`, we should probably throw that immediately so that the server shuts down asap when it can no longer process requests

> I introduced MQEngineBatchError thinking that in the future vLLM would be more robust to those errors, raise an error and keep up and running, but maybe it is a too soon feature

Yeah, it's a nice touch but I don't think we currently have the robustness to use it, like you say any exception will kill the engine :(


This is super close, I think I just have these two comments left then:

1. I'd rather not include the new error type `MQEngineBatchError` since we don't yet have a case where an exception is raised but the engine stays alive. (My care amount on this is low)
2. I want to make sure we're always killing the engine and server ASAP once it errors. (My care amount on this is high). I think this is already mostly covered by your code, we'll just want to make sure that we don't run into the case you mentioned about raising one error first and then waiting for the next request to raise an `MQEngineDeadError`. Maybe a catch-all for this could be like
```
@app.exception_handler(RuntimeError)
async def handler(r, e):
    if engine.errored:
        # log just the exception message
        # shut down server
    else:
        # log full stack trace
```
But I think that's very ham-fisted and probably not necessary 😉